core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -234,6 +234,9 @@ public void run() {
             });
             updateNode(newOne);
             NodeListener.fireOnUpdated(oldOne, newOne);
+            // no need for a full save() so we just do the minimum
+            Util.deleteRecursive(new File(getNodesDir(), oldOne.getNodeName()));
```

Comment:
Do we not need to check here that the old and new nodes have distinct names? If I am reading this right, since you are doing this unconditionally, in the normal case of changing node configuration _without_ renaming you will be deleting the configuration from disk—so that after a restart the node will have disappeared!

Reply:
On line 225 in the condition `if (oldOne == nodes.get(oldOne.getNodeName()))`, this is already checked for. I'm deleting only under this conditional

Reply:
My understanding of the code is that this main branch of the method is run under all normal conditions, whether the node is being renamed or not, and that the `else` branch is only for special cases such as the node having been deleted by another user between the time the configuration page was loaded and the form is saved. In other words, you would need to gate the deletion on something like

```java
if (!newOne.getNodeName().equals(oldOne.getNodeName())) {
```

Regardless, the tests (whether using reload or a true restart) should be checking both a regular save and a rename.

Reply:
Yes, you are right. I'll update the patch.